src/client/texture.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+Minetest
+Copyright (C) 2019 kilbith, Jean-Patrick Guerrero <jeanpatrick.guerrero@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "texture.h"
+#include "porting.h"
+
+void Texture::set(const std::string &_name, const std::string &base_name,
+		s32 _frame_count, u64 _frame_duration)
+{
+	m_name = _name;
+	m_frame_count = _frame_count;
+	m_frame_idx = 0;
+	m_frame_duration = _frame_duration;
+	m_frame_time = 0;
+	m_frame_names.resize(m_frame_count);
+
+	for (s32 i = 0; i < m_frame_count; ++i)
+		m_frame_names[i] = base_name + "_" + (char)(i + 1 + '0') + ".png";
+}
+
+void Texture::set(const std::string &_name)
+{
+	m_name = _name;
+	m_frame_count = 0;
+	m_frame_idx = 0;
+	m_frame_duration = 0;
+	m_frame_time = 0;
+	m_frame_names.resize(0);
+}
+
+std::string Texture::getName() const
+{
+	return m_name;
+}
+
+video::ITexture *Texture::getTexture(ISimpleTextureSource *tsrc)
+{
+	if (m_frame_count > 0)
+		return tsrc->getTexture(m_frame_names[m_frame_idx]);
+	else
+		return tsrc->getTexture(m_name);
+}
+
+void Texture::step(u64 step_duration)
+{
+	m_frame_time += step_duration;
+	m_frame_idx += u32(m_frame_time / m_frame_duration);
+	m_frame_idx %= m_frame_count;
+	m_frame_time %= m_frame_duration;
```

Comment:
Shouldn't this be `m_frame_time %= m_frame_duration * m_frame_count;`?

Reply:
No, it's correct. `m_frame_time` only refers to the current frame here, it's used to advance `m_frame_idx` at an interval of `m_frame_duration` ms.

If we change the line to `m_frame_time %= m_frame_duration * m_frame_count;`, what happens is that the value gets high enough to advance `m_frame_idx` up to `m_frame_count` times per-frame which messes up the playback.